### PR TITLE
Implement LVGL

### LINKED ISSUE #299: Switch to a proper LCD library for the OLED
Espressif has a ESP-IDF library for LCD that's supports the I2C SSD1306 that we use on the Bitaxe. https://docs.espressif.com/projects/esp-idf/en/v5.2.2/esp32s3/api-reference/peripherals/lcd.html

Example for the I2C SSD1306
https://github.com/espressif/esp-idf/tree/v5.2.2/examples/peripherals/lcd/i2c_oled

### PR DESCRIPTION
**Pull request by mutatrum**
This PR adds LVGL 8.4 and refactors all current screens to use it.

The font is from https://int10h.org/oldschool-pc-fonts/fontlist/font?portfolio_6x8, converted to an LVGL font. There are also other fonts available, but currently unused.

The logo bitmap has been re-converted into an LVGL image.

This is a very basic implementation, almost pixel perfect close to the current screens. I think it makes sense to have this running for a while, to see if it's stable before venturing further into fancier usage of displays.

I haven't tested the self-test yet.

Fixes #299

### CODE REVIEW

**Comment by mutatrum** [comment]
Comparison of the flash size, before and after:
```
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Memory Type/Section ┃ Used [bytes] ┃ Used [%] ┃ Remain [bytes] ┃ Total [bytes] ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ Flash Code          │       736098 │     8.78 │        7652478 │       8388576 │
│    .text            │       736098 │     8.78 │                │               │
│ Flash Data          │       192460 │     0.57 │       33361940 │      33554400 │
│    .rodata          │       192204 │     0.57 │                │               │
│    .appdesc         │          256 │      0.0 │                │               │
│ DIRAM               │       123780 │    36.22 │         217980 │        341760 │
│    .text            │        79679 │    23.31 │                │               │
│    .bss             │        21536 │      6.3 │                │               │
│    .data            │        22372 │     6.55 │                │               │
│ IRAM                │        16383 │    99.99 │              1 │         16384 │
│    .text            │        15356 │    93.73 │                │               │
│    .vectors         │         1027 │     6.27 │                │               │
│ RTC FAST            │          284 │     3.47 │           7908 │          8192 │
│    .rtc_reserved    │           24 │     0.29 │                │               │
│    .force_fast      │            4 │     0.05 │                │               │
└─────────────────────┴──────────────┴──────────┴────────────────┴───────────────┘
Total image size: 1046996 bytes (.bin may be padded larger)

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Memory Type/Section ┃ Used [bytes] ┃ Used [%] ┃ Remain [bytes] ┃ Total [bytes] ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ Flash Code          │       836930 │     9.98 │        7551646 │       8388576 │
│    .text            │       836930 │     9.98 │                │               │
│ Flash Data          │       214292 │     0.64 │       33340108 │      33554400 │
│    .rodata          │       214036 │     0.64 │                │               │
│    .appdesc         │          256 │      0.0 │                │               │
│ DIRAM               │       155796 │    45.59 │         185964 │        341760 │
│    .text            │        79679 │    23.31 │                │               │
│    .bss             │        54272 │    15.88 │                │               │
│    .data            │        21652 │     6.34 │                │               │
│ IRAM                │        16383 │    99.99 │              1 │         16384 │
│    .text            │        15356 │    93.73 │                │               │
│    .vectors         │         1027 │     6.27 │                │               │
│ RTC FAST            │          284 │     3.47 │           7908 │          8192 │
│    .rtc_reserved    │           24 │     0.29 │                │               │
│    .force_fast      │            4 │     0.05 │                │               │
└─────────────────────┴──────────────┴──────────┴────────────────┴───────────────┘
Total image size: 1168940 bytes (.bin may be padded larger)
```

**Comment by mutatrum** [comment]
As a bare minimum for leveraging LVGL, I added scrolling labels for long values. The main one would the the pool url, which is not restricted to 20 characters.

**Comment by mutatrum** [comment]
LVGL 9.2 is blocked by https://github.com/espressif/esp-bsp/pull/392.

**Comment by eandersson** [comment]
@mutatrum can you rebase this patch.

**Comment by mutatrum** [comment]
I don't understand why GitHub shows all intermediate commits here. The rebase was clean in my local checkout.

**Comment by mutatrum** [comment]
Ok, it's clean now. Squashed everything into a single commit. Some notes:

 * This overwrites `sdkconfig.defaults` as it was generated by `idf.py save-defconfig`. There were several stale options and some missing options. If you change anything in the menu config, re-run `save-defconfig` to get an updated state, do not manually edit this.
 * This enables the built-in  PSRAM. This is not stricktly neccesary for LVGL, it does run without this enabled but it's running fairly low on memory. I'm not 100% convinced this is a safe thing to do, as I don't know if all the BitAxes ever built use the ESP32S3-WROOM-1-n16r8. If you have a model with different type memory (quad vs octal, or no memory), it'll crash. I don't know if it's possible to auto-detect this.
 * I flipped screen one and two, it now shows IP address and pool as the first screen, as the asic is still ramping up at this time.
 * The font used is: https://int10h.org/oldschool-pc-fonts/fontlist/font?portfolio_6x8, converted with `lv_font_conf` tool. I think this is exactly the same font.

**Comment by mutatrum** [comment]
I opted to do an almost 1-to-1 implementation first to get a clean change-set. The next step would be to inverse the control of displays, and have it refresh values live. This means that a lot of code from `system.c` would move to `display.c`, and the display class pulls data periodically out of the global state. This also make more sense if there are going to be different screen implementations.

File: sdkconfig.defaults
Comment (eandersson) [inline sdkconfig.defaults:11]: You'll need to revert all the changes to sdkconfig.defaults and add your changes on top.

**Review by eandersson — DISMISSED** [review]

**Comment by eandersson** [comment]
There is a crash if the lcd panel isn't installed (or broken) that will need to be fixed.
```
E (1370) display: Install SSD1306 panel driver
E (1370) i2c.master: I2C transaction unexpected nack detected
E (1380) i2c.master: s_i2c_synchronous_transaction(872): I2C transaction failed
E (1380) i2c.master: i2c_master_transmit(1074): I2C transaction failed
E (1390) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(179): i2c transaction failed
E (1400) lcd_panel.ssd1306: panel_ssd1306_init(147): io tx param SSD1306_CMD_SET_MULTIPLEX failed
ESP_ERROR_CHECK failed: esp_err_t 0x103 (ESP_ERR_INVALID_STATE) at 0x420128a3
0x420128a3: display_init at /app/main/display.c:69 (discriminator 1)

file: "./main/display.c" line 69
func: display_init
expression: esp_lcd_panel_init(panel_handle)

abort() was called at PC 0x4037d797 on core 0
0x4037d797: _esp_error_check_failed at /opt/esp/idf/components/esp_system/esp_err.c:49

```

File: sdkconfig.defaults
Comment (mutatrum) [reply sdkconfig.defaults:11]: There were two settings missing indeed:
CONFIG_ESP_DEFAULT_CPU_FREQ_MHZ_240
CONFIG_ESP_WIFI_11KV_SUPPORT

Some fields are removed as they have the default value:
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME
CONFIG_PARTITION_TABLE_FILENAME
CONFIG_PARTITION_TABLE_OFFSET
CONFIG_PARTITION_TABLE_MD5

Some fields are superfluous, as they are already set by a different field:
CONFIG_ESPTOOLPY_FLASHSIZE (set by CONFIG_ESPTOOLPY_FLASHSIZE_16MB)

Some fields have a more general version:
CONFIG_ESP32S3_DEFAULT_CPU_FREQ_240 -> CONFIG_ESP_DEFAULT_CPU_FREQ_MHZ_240

So it's a bit of a clean-up.

The only thing I'm not sure about is the CONFIG_STRATUM_USER and CONFIG_FALLBACK_STRATUM_USER which are added to this file as well if you run save-defconfig. I don't think that should be happening?

I can do a complete revert and add new fields to the end, but I think this is cleaner.

**Comment by mutatrum** [comment]
Without a display, it only fails on the `init` call, which was a bit unexpected. It now continues and starts hashing without a display:
```
I (2203) display: Install panel IO
I (2203) display: Install SSD1306 panel driver
I (2213) lcd_panel.ssd1306: ssd1306 init
E (2223) i2c.master: I2C transaction unexpected nack detected
E (2223) i2c.master: s_i2c_synchronous_transaction(872): I2C transaction failed
E (2233) i2c.master: i2c_master_transmit(1074): I2C transaction failed
E (2243) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(204): i2c transaction failed
E (2253) lcd_panel.ssd1306: panel_ssd1306_init(150): io tx param SSD1306_CMD_SET_MULTIPLEX failed
E (2253) display: display_init(82): Display init failed
W (2263) SystemModule: OLED init failed!
I (2273) SystemModule: SYSTEM_task started
```

File: sdkconfig.defaults
Comment (eandersson) [reply sdkconfig.defaults:11]: I would probably personally prefer to keep to only make the sdkconfig.default changes directly related to the display, but I don't feel strongly about it.

**Comment by eandersson** [comment]
Free Heap Memory changed drastically for me, maybe related to the sdk changes?

**Comment by eandersson** [comment]
Ah yep. If I use the old sdkconfig heap is the same as in previous versions
```
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions.csv"
CONFIG_PARTITION_TABLE_FILENAME="partitions.csv"
CONFIG_PARTITION_TABLE_OFFSET=0x8000
CONFIG_PARTITION_TABLE_MD5=y
CONFIG_ESPTOOLPY_FLASHSIZE_16MB=y
CONFIG_ESPTOOLPY_FLASHSIZE="16MB"
CONFIG_HTTPD_WS_SUPPORT=y
CONFIG_SPIFFS_OBJ_NAME_LEN=64
CONFIG_HTTPD_MAX_URI_LEN=2048
CONFIG_ESP32S3_DEFAULT_CPU_FREQ_240=y
CONFIG_ESP_WIFI_11KV_SUPPORT=y
CONFIG_LV_USE_THEME_DEFAULT=n
CONFIG_LV_USE_THEME_BASIC=n
CONFIG_LV_USE_SNAPSHOT=n
CONFIG_LV_BUILD_EXAMPLES=n
```

**Comment by mutatrum** [comment]
> This enables the built-in  PSRAM. This is not stricktly neccesary for LVGL, it does run without this enabled but it's running fairly low on memory. I'm not 100% convinced this is a safe thing to do, as I don't know if all the BitAxes ever built use the ESP32S3-WROOM-1-n16r8. If you have a model with different type memory (quad vs octal, or no memory), it'll crash. I don't know if it's possible to auto-detect this.

No, it's enabled because the n16r8 has 8MB of PSRAM on board. This is enabled by:
```
CONFIG_SPIRAM=y
CONFIG_SPIRAM_MODE_OCT=y
```

As said above, not 100% sure about this. I think this is the model that's on the BOM, so I would assume everyone picked this variant. But no way to know for sure. I have a few other S3's, but they are all the same variant so I cannot test otherwise.

**Comment by eandersson** [comment]
> > This enables the built-in  PSRAM. This is not stricktly neccesary for LVGL, it does run without this enabled but it's running fairly low on memory. I'm not 100% convinced this is a safe thing to do, as I don't know if all the BitAxes ever built use the ESP32S3-WROOM-1-n16r8. If you have a model with different type memory (quad vs octal, or no memory), it'll crash. I don't know if it's possible to auto-detect this.
> 
> No, it's enabled because the n16r8 has 8MB of PSRAM on board. This is enabled by:
> 
> ```
> CONFIG_SPIRAM=y
> CONFIG_SPIRAM_MODE_OCT=y
> ```
> 
> As said above, not 100% sure about this. I think this is the model that's on the BOM, so I would assume everyone picked this variant. But no way to know for sure. I have a few other S3's, but they are all the same variant so I cannot test otherwise.

Yea - I would def prefer reducing the scope of the changes to make it easier to test etc and you can open a second PR if you want to make additional changes.

**Comment by eandersson** [comment]
Oher than that this PR looks solid now.

**Comment by mutatrum** [comment]
Current free ram after an hour: 92kb bytes.

I'll run it for a while to see if this causes issues.

File: sdkconfig.defaults
Comment (eandersson) [inline sdkconfig.defaults:12]: How about this? Is this default? I still recommend just reverting to the previous to avoid red herring situations.

File: sdkconfig.defaults
Comment (mutatrum) [reply sdkconfig.defaults:11]: I'll extract and make a separate issue for using save-defconfig and cleaning up `sdkconfig.defaults`.

**Comment by mutatrum** [comment]
> Current free ram after an hour: 92k.

Still 92k free ran after 16 hours. Not a lot but stable.

**Comment by mutatrum** [comment]
Probably better to bump this over to at least the next version. I think they'll merge the fix for LVGL 9.2 soon, and I want to move to a minimal `lv_conf.h` so it only compiles LVGL features that are actually used. Which is almost nothing at the moment.

**Comment by mutatrum** [comment]
After rebase on top of 2.4.0, free memory is ~71kb. Binary size is a tad smaller due to disabling all unused LVGL features.

**Comment by skot** [comment]
This looks good to me.. Are we waiting for LVGL 9.2 support to merge?

**Comment by mutatrum** [comment]
Missed your comment, yes, at this point I would wait. Fix on their side needs approval and merge, which should only take a few days.

**Comment by mutatrum** [comment]
`Free Heap Memory: 37612` :eyes:

I'll keep an eye out, otherwise maybe get the PSRAM enabled as well for this?

**Comment by mutatrum** [comment]
`esp_lvgl_port` has been fixed in 2.4.3, so this is now properly on LVGL 9.

**Comment by mutatrum** [comment]
superseded by #539.

**Pull request closed without merging**
